Emit a warning in clickhouse-client if Ordinary database engine is used

shabroo opened the pull request:
### Changelog category (leave one):
- Improvement



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Ordinary database engine is deprecated. You will receive a warning in clickhouse-client if your server is using it. This closes https://github.com/ClickHouse/ClickHouse/issues/52229

Commit:
Hello from Marina

shabroo committed:
Me again! My first attempt at a contribution

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=56942) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=56942) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Marina Fathouat<br/>:x: shabroo<br/><hr/>**Marina Fathouat** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=56942) it.</sub>

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #56942 from shabroo/ClickHouse:master -->
*This is an automated comment for commit 31526d73235ae9a66845f337df77bc317d5eb564 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56942/31526d73235ae9a66845f337df77bc317d5eb564/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6906865426">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56942/31526d73235ae9a66845f337df77bc317d5eb564/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

evillique commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,24 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
     for (const auto & [name, db_path] : databases)
     {
         loadDatabase(context, name, db_path, has_force_restore_data_flag);
         loaded_databases.insert({name, DatabaseCatalog::instance().getDatabase(name)});
+        auto database = DatabaseCatalog::instance().getDatabase(name);
```
```suggestion

        auto database = DatabaseCatalog::instance().getDatabase(name);
        loaded_databases.insert({name, database});
        
```

nikitamikhaylov committed:
Update src/Interpreters/loadMetadata.cpp

Co-authored-by: Nikolay Degterinsky <43110995+evillique@users.noreply.github.com>

alexey-milovidov commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,26 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
+        if (database->getEngineName() == "Ordinary")
+            has_ordinary_databases = true;
+        
+        loaded_databases.insert({name,database});
```
This was already done.

shabroo replied:
OK

alexey-milovidov commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,26 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
     }
 
+    if (has_ordinary_databases)
+        context->addWarningMessage("Server Has Databases with Ordinary Engine which is not optimal. \n " 
```
`Has`, `Databases` - wrong capitalization.

alexey-milovidov replied:
It will be good to tell the name of the first database with the Ordinary engine.

alexey-milovidov commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,26 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
     }
 
+    if (has_ordinary_databases)
+        context->addWarningMessage("Server Has Databases with Ordinary Engine which is not optimal. \n " 
```
Trailing whitespace.
Unneeded whitespace at the beginning of the string.

shabroo replied:
OK, I tried to fix this - not sure if it worked.

alexey-milovidov commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,26 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
 
+    if (has_ordinary_databases)
+        context->addWarningMessage("Server Has Databases with Ordinary Engine which is not optimal. \n " 
+            "To convert this database to a new Atomic engine, please create a flag named `convert_ordinary_to_atomic` in flags directory");
```
What is the flags directory? (needs clarification)

alexey-milovidov commented:
An integration test is needed.

shabroo committed:
Update loadMetadata.cpp

Fixing use of capitals in user message "Server has databases with Ordinary engine, which is not optimal."

tavplubix commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -221,13 +221,26 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
     }
 
+    if (has_ordinary_databases)
+        context->addWarningMessage("Server has databases with Ordinary engine, which is not optimal. \n " 
+            "To convert this database to a new Atomic engine, please create a flag named `convert_ordinary_to_atomic` in flags directory");
```
We can move this line to https://github.com/ClickHouse/ClickHouse/blob/c4fb0a0cd520faeb61ac0cdb9b834062a558b1b4/src/Databases/DatabaseFactory.cpp#L166-L172 or to https://github.com/ClickHouse/ClickHouse/blob/8013cb1f784f6324b3c7b227499751dc7e666009/src/Databases/DatabaseOrdinary.cpp#L81-L84 so we will be able to remove other lines added in this PR, so it will be just one-line change

nikitamikhaylov replied:
With recent changes (when we include the name of the first database which has Ordinary) I think better to keep the code as is.

nikitamikhaylov replied:
Ok I moved the code to the Context.

tavplubix commented:
> An integration test is needed.

It can be a simple stateless functional test. We should prefer functional tests over integration tests when possible

shabroo committed:
Update loadMetadata.cpp

removed space at the beginning of the line

nikitamikhaylov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into shabroo/master

nikitamikhaylov committed:
Better

nikitamikhaylov committed:
Better

nikitamikhaylov committed:
Update 02988_ordinary_database_warning.sql

nikitamikhaylov committed:
Update tests

nikitamikhaylov committed:
Revert "Update tests"

This reverts commit 06ef0d23b7e047e78f19d5ff0dc832a8f4b21949.

nikitamikhaylov committed:
Better

nikitamikhaylov committed:
Fix test

nikitamikhaylov committed:
Update 02021_create_database_with_comment.sh

nikitamikhaylov committed:
Fix test

nikitamikhaylov committed:
Maybe better?

nikitamikhaylov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into shabroo/master

nikitamikhaylov committed:
Im tired

nikitamikhaylov committed:
Another try

nikitamikhaylov committed:
What else

nikitamikhaylov committed:
Update tests

nikitamikhaylov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into shabroo/master

nikitamikhaylov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into shabroo/master

nikitamikhaylov commented:
Functional tests - all failing in master, will take a closer look after merging. 
AST Fuzzer - the fuzzer itself crashed and we don't have enough info to understand the reason. However I see the same problem reported from master.
